Description: Fix the build on non-Linux architectures.
 - never build with the "Linux eccentricities" on Debian
 - make the CMSG_CONTROLDATA() macro work and introduce
   an lvalue equivalent
 - add two more kFreeBSD tests in the configure script
Author: Peter Pentchev <roam@ringlet.net>
Forwarded: no
Last-Update: 2016-04-06

--- a/include/common.h
+++ b/include/common.h
@@ -1088,7 +1088,8 @@
  * cast is necessary on AIX, due to buggy headers there?
  * needs additional testing on AIX, disable for now.
  */
-#define CMSG_CONTROLDATA(msg)   ((msg).msg_control)
+#define CMSG_CONTROLDATA(msg)   ((struct cmsghdr *)((msg).msg_control))
+#define CMSG_CONTROLDATA_MUTABLE(msg)   ((msg).msg_control)
 #else /* !HAVE_CMSGHDR */
 #define CMSG_CONTROLDATA(msg)   ((msg).msg_accrights)
 #endif /* HAVE_CMSGHDR */
--- a/lib/Rcompat.c
+++ b/lib/Rcompat.c
@@ -293,7 +293,7 @@
 
    /* no cmsg on proxied sockets. */
    CMSG_TOTLEN(*msg)      = 0;
-   CMSG_CONTROLDATA(*msg) = NULL;
+   CMSG_CONTROLDATA_MUTABLE(*msg) = NULL;
 
    for (received = ioc = rc = 0; ioc < (size_t)msg->msg_iovlen; ++ioc) {
       if ((rc = Rrecvfrom(s,
--- a/preload.m4
+++ b/preload.m4
@@ -438,7 +438,7 @@
     AC_DEFINE(HAVE_PTHREAD_H, 1, [have pthread header])
     if test x"${ac_cv_search_pthread_mutexattr_init}" = x"-lpthread"; then
         case $host in
-	    *-*-linux-*)
+	    *-*-linux-*|*-gnu)
 	        #XXX attempt to find latest pthread library
 		PATH=$PATH:/sbin
 		export PATH
@@ -616,7 +616,7 @@
 	    fi
 	    ;;
 
-	*-*-linux-*)
+	*-*-linux-*|*-gnu)
 	    #XXX
 	    #can't set it to libc.so directly, might be ld script
 	    unset LIBC_ALTS
